packages/explorer/src/app/(explorer)/[chainName]/worlds/[worldAddress]/decode/DecodeForm.tsx

```diff
@@ -43,12 +56,50 @@ function isAbiError(item: AbiItem): item is AbiItem & { type: "error" } {
 export function DecodeForm() {
   const { data: worldData, isLoading: isWorldAbiLoading } = useWorldAbiQuery();
   const { data: systemData, isLoading: isSystemAbisLoading } = useSystemAbisQuery();
-  const [abiItem, setAbiItem] = useState<AbiFunction | AbiError>();
+  const [decoded, setDecoded] = useState<AbiFunction | AbiError | ResourceItem | SelectorDatabase>();
+  const [encoded, setEncoded] = useState<string | undefined>();
   const form = useForm<z.infer<typeof formSchema>>({
     resolver: zodResolver(formSchema),
   });
 
-  function onSubmit({ selector }: z.infer<typeof formSchema>) {
+  const results = useMemo<Result[]>(() => {
+    if (!decoded) return [];
+    if (decoded.type === "resource") {
+      return [{ type: decoded.resource.type, label: resourceToLabel(decoded.resource) }];
+    }
+    if (decoded.type === "function" || decoded.type === "error") {
+      return [{ type: decoded.type, label: formatAbiItem(decoded) }];
+    }
+    if (decoded.type === "signature") {
+      return decoded.selectors.map((selector) => ({ type: "signature", label: selector }));
+    }
+    return [];
+  }, [decoded]);
+
+  const decodedFunctionCalls = useMemo(() => {
+    try {
+      if (!decoded || !encoded) return;
+      if (decoded.type === "function") {
+        return [{ ...decodeFunctionData({ abi: [decoded], data: encoded as Hex }), abi: decoded }];
+      }
+      if (decoded.type === "error") {
+        return [{ ...decodeErrorResult({ abi: [decoded], data: encoded as Hex }), abi: decoded }];
+      }
+      if (decoded.type === "signature") {
+        return decoded.selectors.map((selector) => {
+          const abiItem = parseAbiItem(`function ${selector}`) as AbiFunction;
+          return { ...decodeFunctionData({ abi: [abiItem], data: encoded as Hex }), abi: abiItem };
+        });
+      }
+    } catch (error) {
+      console.error("Error decoding function data", error);
+    }
+  }, [decoded, encoded]);
+
```

Comment:
Not sure I completely dig the useMemo approach, thinking if handling decoding onSubmit would also work and require one less useState, like it was earlier. But it's also fine, I guess it's somewhat a personal preference.